sound/soc/codecs/rt715-sdca.c

```diff
@@ -0,0 +1,896 @@
+			ctl_r = ~(val_r - 1) & 0xffff;
+			neg_flag = 1;
+		}
+		ctl_r *= 1000;
+		ctl_r >>= 8;
+		if (neg_flag)
+			ctl_r = mc->shift - ctl_r / 375;
+		else
+			ctl_r = mc->shift + ctl_r / 375;
+	}
+
+	ucontrol->value.integer.value[0] = ctl_l;
+	ucontrol->value.integer.value[1] = ctl_r;
+
+	return 0;
+}
+
+static const DECLARE_TLV_DB_SCALE(in_vol_tlv, -17625, 375, 0);
+static const DECLARE_TLV_DB_SCALE(mic_vol_tlv, 0, 1000, 0);
+
+#define SOC_DOUBLE_R_EXT(xname, reg_left, reg_right, xshift, xmax, xinvert,\
+	 xhandler_get, xhandler_put) \
+{	.iface = SNDRV_CTL_ELEM_IFACE_MIXER, .name = (xname), \
+	.info = snd_soc_info_volsw, \
+	.get = xhandler_get, .put = xhandler_put, \
+	.private_value = SOC_DOUBLE_R_VALUE(reg_left, reg_right, xshift, \
+					    xmax, xinvert) }
+
+static const struct snd_kcontrol_new rt715_snd_controls_sdca[] = {
+	/* Capture switch */
+	SOC_DOUBLE_R("ADC 07_27 Capture Switch",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC7_27_VOL, FU_MUTE_CTRL, CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC7_27_VOL, FU_MUTE_CTRL, CH_02),
+			0, 1, 1),
+	SOC_DOUBLE_R("ADC 09 Capture Switch",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_MUTE_CTRL, CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_MUTE_CTRL, CH_02),
+			0, 1, 1),
+	SOC_DOUBLE_R("ADC 08 Capture Switch",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_MUTE_CTRL, CH_03),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_MUTE_CTRL, CH_04),
+			0, 1, 1),
+	SOC_DOUBLE_R("ADC 10 Capture Switch",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_MUTE_CTRL, CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_MUTE_CTRL, CH_02),
+			0, 1, 1),
+	SOC_DOUBLE_R("ADC 11 Capture Switch",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_MUTE_CTRL, CH_03),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_MUTE_CTRL, CH_04),
+			0, 1, 1),
+	/* Volume Control */
+	SOC_DOUBLE_R_EXT_TLV("ADC 07_27 Capture Volume",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC7_27_VOL, FU_VOL_CTRL, CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC7_27_VOL, FU_VOL_CTRL, CH_02),
+			0x2f, 0x7f, 0,
+		rt715_set_amp_gain_get_sdca, rt715_set_amp_gain_put_sdca,
+		in_vol_tlv),
+	SOC_DOUBLE_R_EXT_TLV("ADC 09 Capture Volume",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_VOL_CTRL, CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_VOL_CTRL, CH_02),
+			0x2f, 0x7f, 0,
+		rt715_set_amp_gain_get_sdca, rt715_set_amp_gain_put_sdca,
+		in_vol_tlv),
+	SOC_DOUBLE_R_EXT_TLV("ADC 08 Capture Volume",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_VOL_CTRL,
+			CH_03),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC8_9_VOL, FU_VOL_CTRL,
+			CH_04), 0x2f, 0x7f, 0,
+		rt715_set_amp_gain_get_sdca, rt715_set_amp_gain_put_sdca,
+		in_vol_tlv),
+	SOC_DOUBLE_R_EXT_TLV("ADC 10 Capture Volume",
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_VOL_CTRL,
+			CH_01),
+		RT715_SDCA_ADDR(FUNC_02, FU_ADC10_11_VOL, FU_VOL_CTRL,
+			CH_02), 0x2f, 0x7f, 0,
+		rt715_set_amp_gain_get_sdca, rt715_set_amp_gain_put_sdca,
+		in_vol_tlv),
+	SOC_DOUBLE_R_EXT_TLV("ADC 11 Capture Volume",
```

Comment:
I think it'd be good to have a consistent naming of controls across drivers. for rt711 the control name includes more information that makes it less ambiguous. Something to discuss between @shumingfan and @jack-cy-yu